pvlib/modelchain.py

```diff
@@ -1090,39 +1248,83 @@ def _verify_df(self, data, required):
+        if not isinstance(data, tuple):
+            _verify(data)
+        else:
+            for array_data in data:
+                _verify(array_data)
 
     def _assign_weather(self, data):
-        key_list = [k for k in WEATHER_KEYS if k in data]
-        self.weather = data[key_list].copy()
-        if self.weather.get('wind_speed') is None:
-            self.weather['wind_speed'] = 0
-        if self.weather.get('temp_air') is None:
-            self.weather['temp_air'] = 20
+        def _build_weather(data):
+            key_list = [k for k in WEATHER_KEYS if k in data]
+            weather = data[key_list].copy()
+            if weather.get('wind_speed') is None:
+                weather['wind_speed'] = 0
+            if weather.get('temp_air') is None:
+                weather['temp_air'] = 20
+            return weather
+        if not isinstance(data, tuple):
+            self.weather = _build_weather(data)
+        else:
+            self.weather = tuple(
+                _build_weather(weather) for weather in data
+            )
         return self
 
     def _assign_total_irrad(self, data):
-        key_list = [k for k in POA_KEYS if k in data]
-        self.total_irrad = data[key_list].copy()
+        def _build_irrad(data):
+            key_list = [k for k in POA_KEYS if k in data]
+            return data[key_list].copy()
+        if isinstance(data, tuple):
+            self.results.total_irrad = tuple(
+                _build_irrad(irrad_data) for irrad_data in data
+            )
+            return self
+        self.results.total_irrad = _build_irrad(data)
         return self
 
+    def _assign_times(self):
+        """Assign self.times according the the index of self.weather.
+
+        If there are multiple DataFrames in self.weather then the index
+        of the first one is assigned
+        """
+        if isinstance(self.weather, tuple):
+            self.times = self.weather[0].index
+        else:
+            self.times = self.weather.index
+
     def prepare_inputs(self, weather):
         """
         Prepare the solar position, irradiance, and weather inputs to
         the model, starting with GHI, DNI and DHI.
 
         Parameters
         ----------
-        weather : DataFrame
+        weather : DataFrame or tuple of DataFrame
             Column names must be ``'dni'``, ``'ghi'``, ``'dhi'``,
             ``'wind_speed'``, ``'temp_air'``. All irradiance components
             are required. Air temperature of 20 C and wind speed
             of 0 m/s will be added to the DataFrame if not provided.
 
+            If `weather` is a tuple each DataFrame it contains must have
+            the same index and it must be the same length as the number
+            of Arrays in the system.
+
+        Raises
+        ------
+        ValueError
+            If the `weather` DataFrame(s) are missing an irradiance component,
+            if `weather` is a tuple and the DataFrames it contains do not
```

Comment:
I usually make separate entries for each of these possible ValueErrors (as shown below), but I don't actually know if that's the generally recommended practice.

```
Raises
------
ValueError
    Reason 1
ValueError
    Reason 2
```

Reply:
That seems substantially easier to read "in anger" than a single dense sentence.